// aws-iot-device-sdk-v2/dist/iotidentity/v2utils.js
"use strict";
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0.
 *
 * This file is generated
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeServiceModel = void 0;
const util_utf8_browser_1 = require("@aws-sdk/util-utf8-browser");
const model_validation_utils = __importStar(require("../mqtt_request_response_utils"));
function normalizeCreateCertificateFromCsrRequest(value) {
    let normalizedValue = {};
    if (value.certificateSigningRequest) {
        normalizedValue.certificateSigningRequest = value.certificateSigningRequest;
    }
    return normalizedValue;
}
function buildCreateCertificateFromCsrRequestPayload(request) {
    let value = normalizeCreateCertificateFromCsrRequest(request);
    return (0, util_utf8_browser_1.fromUtf8)(JSON.stringify(value));
}
function applyCorrelationTokenToCreateCertificateFromCsrRequest(request) {
    return [request, undefined];
}
function normalizeCreateKeysAndCertificateRequest(value) {
    let normalizedValue = {};
    return normalizedValue;
}
function buildCreateKeysAndCertificateRequestPayload(request) {
    let value = normalizeCreateKeysAndCertificateRequest(request);
    return (0, util_utf8_browser_1.fromUtf8)(JSON.stringify(value));
}
function applyCorrelationTokenToCreateKeysAndCertificateRequest(request) {
    return [request, undefined];
}
function normalizeRegisterThingRequest(value) {
    let normalizedValue = {};
    if (value.certificateOwnershipToken) {
        normalizedValue.certificateOwnershipToken = value.certificateOwnershipToken;
    }
    if (value.parameters) {
        normalizedValue.parameters = value.parameters;
    }
    return normalizedValue;
}
function buildRegisterThingRequestPayload(request) {
    let value = normalizeRegisterThingRequest(request);
    return (0, util_utf8_browser_1.fromUtf8)(JSON.stringify(value));
}
function applyCorrelationTokenToRegisterThingRequest(request) {
    return [request, undefined];
}
function buildCreateCertificateFromCsrSubscriptions(request) {
    return new Array(`$aws/certificates/create-from-csr/json/accepted`, `$aws/certificates/create-from-csr/json/rejected`);
}
function buildCreateCertificateFromCsrPublishTopic(request) {
    return `$aws/certificates/create-from-csr/json`;
}
function buildCreateCertificateFromCsrResponsePaths(request) {
    return new Array({
        topic: `$aws/certificates/create-from-csr/json/accepted`,
        deserializer: deserializeCreateCertificateFromCsrResponse,
    }, {
        topic: `$aws/certificates/create-from-csr/json/rejected`,
        deserializer: deserializeV2ErrorResponse,
    });
}
function buildCreateKeysAndCertificateSubscriptions(request) {
    return new Array(`$aws/certificates/create/json/accepted`, `$aws/certificates/create/json/rejected`);
}
function buildCreateKeysAndCertificatePublishTopic(request) {
    return `$aws/certificates/create/json`;
}
function buildCreateKeysAndCertificateResponsePaths(request) {
    return new Array({
        topic: `$aws/certificates/create/json/accepted`,
        deserializer: deserializeCreateKeysAndCertificateResponse,
    }, {
        topic: `$aws/certificates/create/json/rejected`,
        deserializer: deserializeV2ErrorResponse,
    });
}
function buildRegisterThingSubscriptions(request) {
    let typedRequest = request;
    return new Array(`$aws/provisioning-templates/${typedRequest.templateName}/provision/json/accepted`, `$aws/provisioning-templates/${typedRequest.templateName}/provision/json/rejected`);
}
function buildRegisterThingPublishTopic(request) {
    let typedRequest = request;
    return `$aws/provisioning-templates/${typedRequest.templateName}/provision/json`;
}
function buildRegisterThingResponsePaths(request) {
    let typedRequest = request;
    return new Array({
        topic: `$aws/provisioning-templates/${typedRequest.templateName}/provision/json/accepted`,
        deserializer: deserializeRegisterThingResponse,
    }, {
        topic: `$aws/provisioning-templates/${typedRequest.templateName}/provision/json/rejected`,
        deserializer: deserializeV2ErrorResponse,
    });
}
function deserializeCreateCertificateFromCsrResponse(payload) {
    const payload_text = (0, util_utf8_browser_1.toUtf8)(new Uint8Array(payload));
    return JSON.parse(payload_text);
}
function deserializeCreateKeysAndCertificateResponse(payload) {
    const payload_text = (0, util_utf8_browser_1.toUtf8)(new Uint8Array(payload));
    return JSON.parse(payload_text);
}
function deserializeRegisterThingResponse(payload) {
    const payload_text = (0, util_utf8_browser_1.toUtf8)(new Uint8Array(payload));
    return JSON.parse(payload_text);
}
function deserializeV2ErrorResponse(payload) {
    const payload_text = (0, util_utf8_browser_1.toUtf8)(new Uint8Array(payload));
    return JSON.parse(payload_text);
}
function createRequestResponseOperationServiceModelMap() {
    return new Map([
        ["createCertificateFromCsr", {
                inputShapeName: "CreateCertificateFromCsrRequest",
                payloadTransformer: buildCreateCertificateFromCsrRequestPayload,
                subscriptionGenerator: buildCreateCertificateFromCsrSubscriptions,
                responsePathGenerator: buildCreateCertificateFromCsrResponsePaths,
                publishTopicGenerator: buildCreateCertificateFromCsrPublishTopic,
                correlationTokenApplicator: applyCorrelationTokenToCreateCertificateFromCsrRequest,
            }],
        ["createKeysAndCertificate", {
                inputShapeName: "CreateKeysAndCertificateRequest",
                payloadTransformer: buildCreateKeysAndCertificateRequestPayload,
                subscriptionGenerator: buildCreateKeysAndCertificateSubscriptions,
                responsePathGenerator: buildCreateKeysAndCertificateResponsePaths,
                publishTopicGenerator: buildCreateKeysAndCertificatePublishTopic,
                correlationTokenApplicator: applyCorrelationTokenToCreateKeysAndCertificateRequest,
            }],
        ["registerThing", {
                inputShapeName: "RegisterThingRequest",
                payloadTransformer: buildRegisterThingRequestPayload,
                subscriptionGenerator: buildRegisterThingSubscriptions,
                responsePathGenerator: buildRegisterThingResponsePaths,
                publishTopicGenerator: buildRegisterThingPublishTopic,
                correlationTokenApplicator: applyCorrelationTokenToRegisterThingRequest,
            }],
    ]);
}
function createStreamingOperationServiceModelMap() {
    return new Map([]);
}
function validateCreateCertificateFromCsrRequest(value) {
    // @ts-ignore
    let typedValue = value;
    model_validation_utils.validateValueAsString(value.certificateSigningRequest, 'certificateSigningRequest');
}
function validateCreateKeysAndCertificateRequest(value) {
    // @ts-ignore
    let typedValue = value;
}
function validateRegisterThingRequest(value) {
    // @ts-ignore
    let typedValue = value;
    model_validation_utils.validateValueAsTopicSegment(value.templateName, 'templateName');
    model_validation_utils.validateValueAsString(value.certificateOwnershipToken, 'certificateOwnershipToken');
    model_validation_utils.validateValueAsOptionalMap(value.parameters, model_validation_utils.validateValueAsString, model_validation_utils.validateValueAsString, 'parameters');
}
function createValidatorMap() {
    return new Map([
        ["CreateCertificateFromCsrRequest", validateCreateCertificateFromCsrRequest],
        ["CreateKeysAndCertificateRequest", validateCreateKeysAndCertificateRequest],
        ["RegisterThingRequest", validateRegisterThingRequest],
    ]);
}
function makeServiceModel() {
    let model = {
        requestResponseOperations: createRequestResponseOperationServiceModelMap(),
        streamingOperations: createStreamingOperationServiceModelMap(),
        shapeValidators: createValidatorMap()
    };
    return model;
}
exports.makeServiceModel = makeServiceModel;
